# Feature Selection Techniques


In [4]:
import pandas as pd
import numpy as np

In [5]:
data = pd.read_csv("../../data/50_Startups.csv")

In [6]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     str    
 4   Profit           50 non-null     float64
dtypes: float64(4), str(1)
memory usage: 2.5 KB


In [7]:
data.head()

,R&D Spend,Administration,Marketing Spend,State,Profit
0,165349.20,136897.80,471784.10,New York,192261.83
1,162597.70,151377.59,443898.53,California,191792.06
2,153441.51,101145.55,407934.54,Florida,191050.39
3,144372.41,118671.85,383199.62,New York,182901.99
4,142107.34,91391.77,366168.42,Florida,166187.94


In [8]:
finalDataset = pd.concat([pd.get_dummies(data.State, dtype=int), data.iloc[:, [0,1,2,4]]], axis=1)
finalDataset.info()

<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   California       50 non-null     int64  
 1   Florida          50 non-null     int64  
 2   New York         50 non-null     int64  
 3   R&D Spend        50 non-null     float64
 4   Administration   50 non-null     float64
 5   Marketing Spend  50 non-null     float64
 6   Profit           50 non-null     float64
dtypes: float64(4), int64(3)
memory usage: 2.9 KB


In [9]:
finalDataset.head()

,California,Florida,New York,R&D Spend,Administration,Marketing Spend,Profit
0,0,0,1,165349.20,136897.80,471784.10,192261.83
1,1,0,0,162597.70,151377.59,443898.53,191792.06
2,0,1,0,153441.51,101145.55,407934.54,191050.39
3,0,0,1,144372.41,118671.85,383199.62,182901.99
4,0,1,0,142107.34,91391.77,366168.42,166187.94


###  1. Correlation Analysis: For Regression, we select the features based on corr value w.r.t label


In [10]:
finalDataset.corr()

,California,Florida,New York,R&D Spend,Administration,Marketing Spend,Profit
California,1.000000,-0.492366,-0.515152,-0.143165,-0.015478,-0.168875,-0.145837
Florida,-0.492366,1.000000,-0.492366,0.105711,0.010493,0.205685,0.116244
New York,-0.515152,-0.492366,1.000000,0.039068,0.005145,-0.033670,0.031368
R&D Spend,-0.143165,0.105711,0.039068,1.000000,0.241955,0.724248,0.972900
Administration,-0.015478,0.010493,0.005145,0.241955,1.000000,-0.032154,0.200717
Marketing Spend,-0.168875,0.205685,-0.033670,0.724248,-0.032154,1.000000,0.747766
Profit,-0.145837,0.116244,0.031368,0.972900,0.200717,0.747766,1.000000


### Guideline:
#### Select those feature columns that have corr greater than equal to 50%
#### Selected Features: RDSpend and MarketingSpend

# Backward Elimination using OLS

In [11]:
features = finalDataset.iloc[:, :-1].values
labels = finalDataset.iloc[:, [-1]].values

In [12]:
features

array([[0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.6534920e+05,
        1.3689780e+05, 4.7178410e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.6259770e+05,
        1.5137759e+05, 4.4389853e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.5344151e+05,
        1.0114555e+05, 4.0793454e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.4437241e+05,
        1.1867185e+05, 3.8319962e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.4210734e+05,
        9.1391770e+04, 3.6616842e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.3187690e+05,
        9.9814710e+04, 3.6286136e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.3461546e+05,
        1.4719887e+05, 1.2771682e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.3029813e+05,
        1.4553006e+05, 3.2387668e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.2054252e+05,
        1.4871895e+05, 3.1161329e+05],
       [1.0000000e+00, 0.0000000e+00,

In [13]:
labels

array([[192261.83],
       [191792.06],
       [191050.39],
       [182901.99],
       [166187.94],
       [156991.12],
       [156122.51],
       [155752.6 ],
       [152211.77],
       [149759.96],
       [146121.95],
       [144259.4 ],
       [141585.52],
       [134307.35],
       [132602.65],
       [129917.04],
       [126992.93],
       [125370.37],
       [124266.9 ],
       [122776.86],
       [118474.03],
       [111313.02],
       [110352.25],
       [108733.99],
       [108552.04],
       [107404.34],
       [105733.54],
       [105008.31],
       [103282.38],
       [101004.64],
       [ 99937.59],
       [ 97483.56],
       [ 97427.84],
       [ 96778.92],
       [ 96712.8 ],
       [ 96479.51],
       [ 90708.19],
       [ 89949.14],
       [ 81229.06],
       [ 81005.76],
       [ 78239.91],
       [ 77798.83],
       [ 71498.49],
       [ 69758.98],
       [ 65200.33],
       [ 64926.08],
       [ 49490.75],
       [ 42559.73],
       [ 35673.41],
       [ 14681.4 ]])

In [18]:
np.ones(((len(features)), 1)).astype(int)

array([[1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1]])

#### Step1: Perform All-In

In [19]:
featuresAllIn = np.append( np.ones(((len(features)), 1)).astype(int), features, axis=1 )
featuresAllIn

array([[1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.6534920e+05, 1.3689780e+05, 4.7178410e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 0.0000000e+00,
        1.6259770e+05, 1.5137759e+05, 4.4389853e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.5344151e+05, 1.0114555e+05, 4.0793454e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.4437241e+05, 1.1867185e+05, 3.8319962e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.4210734e+05, 9.1391770e+04, 3.6616842e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.3187690e+05, 9.9814710e+04, 3.6286136e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 0.0000000e+00,
        1.3461546e+05, 1.4719887e+05, 1.2771682e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.3029813e+05, 1.4553006e+05, 3.2387668e+05],
       [1.0000000e+00, 0

#### Step1: Perform All-In

In [21]:
SL=0.05

#### Step3: Perform OLS
endog --- label column
exog ---- feature column

In [22]:
import statsmodels.regression.linear_model as stat

model = stat.OLS(endog=labels, exog=featuresAllIn).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     169.9
Date:                Fri, 26 Jun 2026   Prob (F-statistic):           1.34e-27
Time:                        16:43:09   Log-Likelihood:                -525.38
No. Observations:                  50   AIC:                             1063.
Df Residuals:                      44   BIC:                             1074.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.763e+04   5073.636      7.417      0.000    2.74e+04    4.79e+04
x1          1.249e+04   2449.797      5.099      0.000    7554.868    1.74e+04
x2          1.269e+04   2726.700      4.654      0.000    7195.596    1.82e+04
x3          1.245e+04   2486.364      5.007      0.000    7439.285    1.75e+04
x4             0.8060      0.046     17.369      0.000       0.712       0.900
x5            -0.0270      0.052     -0.517      0.608      -0.132       0.078
x6             0.0270      0.017      1.574      0.123      -0.008       0.062
==============================================================================
Omnibus:                       14.782   Durbin-Watson:                   1.283
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.266
Skew:                          -0.948   Prob(JB):                     2.41e-05
Kurtosis:                       5.572   Cond. No.                     9.69e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 4.13e-24. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#### Step4: Select the feature that has the highest p value

We selected x5 (Admin) === p => 0.608

##### Step5: if pvalue>SL : eliminate feature and recreate new feature set else goto step
Here we are eliminating Admin Column
 updated feature column is

In [23]:
newFeatureColumn = featuresAllIn[:, [1,2,3,4,6]]

#### Step3: Perform OLS
endog --- label column
exog ---- feature column

In [24]:
import statsmodels.regression.linear_model as stat

model = stat.OLS(endog=labels, exog=newFeatureColumn).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.950
Model:                            OLS   Adj. R-squared:                  0.946
Method:                 Least Squares   F-statistic:                     215.8
Date:                Fri, 26 Jun 2026   Prob (F-statistic):           9.72e-29
Time:                        16:45:41   Log-Likelihood:                -525.53
No. Observations:                  50   AIC:                             1061.
Df Residuals:                      45   BIC:                             1071.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1          4.696e+04   3119.471     15.053      0.000    4.07e+04    5.32e+04
x2           4.71e+04   3670.129     12.833      0.000    3.97e+04    5.45e+04
x3          4.694e+04   3342.591     14.043      0.000    4.02e+04    5.37e+04
x4             0.7967      0.042     18.771      0.000       0.711       0.882
x5             0.0298      0.016      1.842      0.072      -0.003       0.062
==============================================================================
Omnibus:                       14.640   Durbin-Watson:                   1.257
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.037
Skew:                          -0.938   Prob(JB):                     2.70e-05
Kurtosis:                       5.565   Cond. No.                     9.45e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.45e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Step4: Select the feature that has the highest p value
We selected x5 (Marketing) === p -> 0.072

In [25]:
newFeatureCol = featuresAllIn[:,[0,1,2,3,4]]

In [26]:
model = stat.OLS(endog=labels, exog=newFeatureCol).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.943
Method:                 Least Squares   F-statistic:                     272.4
Date:                Fri, 26 Jun 2026   Prob (F-statistic):           2.76e-29
Time:                        16:46:40   Log-Likelihood:                -527.35
No. Observations:                  50   AIC:                             1063.
Df Residuals:                      46   BIC:                             1070.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.686e+04   1959.786     18.806      0.000    3.29e+04    4.08e+04
x1          1.189e+04   1956.677      6.079      0.000    7955.697    1.58e+04
x2          1.306e+04   2122.665      6.152      0.000    8785.448    1.73e+04
x3           1.19e+04   2036.022      5.847      0.000    7805.580     1.6e+04
x4             0.8530      0.030     28.226      0.000       0.792       0.914
==============================================================================
Omnibus:                       13.418   Durbin-Watson:                   1.122
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               17.605
Skew:                          -0.907   Prob(JB):                     0.000150
Kurtosis:                       5.271   Cond. No.                     2.48e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.1e-24. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [27]:
newFeatureCol = featuresAllIn[:,[0,1,2,3,4]]

In [28]:
newFeatureCol

array([[1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.6534920e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 0.0000000e+00,
        1.6259770e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.5344151e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.4437241e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.4210734e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.3187690e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 0.0000000e+00,
        1.3461546e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.3029813e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.0000000e+00,
        1.2054252e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 0.0000000e+00,
        1.2333488e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 0.0000000e+00,
        1.01

# Using RFE (Recursive Feature Elimination)

In [29]:
#Step1: Initialize ML Algo
from sklearn.linear_model import LinearRegression
modelAlgo = LinearRegression()

In [30]:
#Step2: Initialize RFE
#step refers to how many feature cols to eliminate per iteration

from sklearn.feature_selection import RFE
selectFeatures = RFE(estimator=modelAlgo, step = 1)

In [31]:
selectFeatures.fit(features, labels)

RFE(estimator=LinearRegression())

In [32]:
selectFeatures.ranking_

array([1, 1, 1, 2, 3, 4])

In [33]:
#Guideline: Select the columns with ranking 1 and 2
#RFE selected 'California', 'Florida', 'New York', 'R&D Spend'


# Select From Model


In [ ]:
#Step1: Initialize ML algo
from sklearn.linear_model import LinearRegression
modelAlgo = LinearRegression()


In [34]:
#Step2: Initialize SFM

from sklearn.feature_selection import SelectFromModel

selectFeatures = SelectFromModel(estimator=modelAlgo)

In [35]:
#Step3: Fit the data to the SFM
selectFeatures.fit(features, labels)

SelectFromModel(estimator=LinearRegression())

In [37]:
finalDataset.columns

Index(['California', 'Florida', 'New York', 'R&D Spend', 'Administration',
       'Marketing Spend', 'Profit'],
      dtype='str')

In [36]:
#Step4: Identify the feature that has highest ranking

selectFeatures.get_support()

array([ True,  True,  True, False, False, False])